source/opt/eliminate_dead_output_stores_pass.cpp

```diff
@@ -0,0 +1,418 @@
+// Copyright (c) 2022 The Khronos Group Inc.
+// Copyright (c) 2022 LunarG Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/opt/eliminate_dead_output_stores_pass.h"
+
+#include "source/opt/instruction.h"
+#include "source/opt/ir_context.h"
+
+namespace {
+
+const uint32_t kDecorationLocationInIdx = 2;
+const uint32_t kTypePointerTypeInIdx = 1;
+const uint32_t kOpTypeVectorTypeInIdx = 0;
+const uint32_t kOpTypeVectorCountInIdx = 1;
+const uint32_t kEntryPointExecutionModelInIdx = 0;
+const uint32_t kOpDecorateMemberMemberInIdx = 1;
+const uint32_t kOpDecorateMemberLocationInIdx = 3;
+
+}  // namespace
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status EliminateDeadOutputStoresPass::Process() {
+  // Current functionality assumes shader capability
+  if (!context()->get_feature_mgr()->HasCapability(SpvCapabilityShader))
+    return Status::SuccessWithoutChange;
+  Pass::Status status;
+  if (analyze_)
+    status = DoDeadOutputStoreAnalysis();
+  else
+    status = DoDeadOutputStoreElimination();
```

Comment:
More than one third of the lines in this file (AnalyzeAccessChain and its support routines) are shared between DoDeadOutputStoreAnalysis and DoDeadOutputStoreElimination.

I could create a LocationAnalysis class and then derive the two passes off of that class. I just thought it would be simpler to do things this way and I didn't see any downside.

Turning the DeadOutputStore analysis here into a global, free standing analysis like the DefUseMgr seems like overkill. There is only one pass that currently cares about it. Why advertise to all the other passes and make them maintain it?